File: javascript/linkedLists.js

class Node {
    constructor(val) {
        this.val = val
        this.next = null
    }
}

class SinglyLinkedList {
    constructor() {
        this.length = 0
        this.head = null
        this.tail = null
    }

    push(val) {
        let node = new Node(val)
        if (this.length === 0) {
            this.head = node
            this.tail = node
        } else {
            this.tail.next = node
            this.tail = node
        }
        this.length ++
        return this
    }
    pop() {
        if (this.length === 0) {
            return null
        } else if (this.length === 1) {
            let temp = this.head
            this.head = null
            this.tail = null
            this.length --
            return temp
        }
        let node = this.head
        for (let i = 0; i < this.length - 2; i++) {
            node = node.next
        }
        let temp = this.tail 
        this.tail = node
        this.tail.next = null
        this.length --
        return temp
    }
    shift() {
        if (this.length === 0) return undefined
        if (this.length === 1) {
            let temp = this.head
            this.head = null
            this.tail = null
            return temp
        }
        let temp = this.head
        this.head = this.head.next
        this.length --
        return temp
    }
    unshift(val) {
        let node = new Node(val)
        if (this.length === 0) {
            this.head = node
            this.tail = node
        } else {
            node.next = this.head
            this.head = node
        }
        this.length ++
        return this
    }
    get(index) {
        if (index < 0 || index > this.length) {
            return null
        }
        let node = this.head
        while(index != 0) {
            node = node.next
            index --
        }
        return node
    }
    set(index, val) {
        let node = this.get(index)
        if (node) {
            node.val = val
            return true
        }
        return false
    }
    insert(index, val) {
        if (index < 0 || index > this.length) {
            return false
        } else if (index === 0) {
            this.unshift(val)
        } else if (index === this.length) {
            this.push(val)
        } else {
            let node = this.get(index - 1)
            let newNode = new Node(val)
            newNode.next = node.next
            node.next = newNode
            this.length ++
        }
        return true
    }
    remove(index) {
        if (index < 0 || index >= this.length) {
            return null
        } else if (index === 0) {
            return this.shift()
        } else if (index === this.length - 1) {
            return this.pop()
        } else {
            let prev = this.get(index - 1)
            let temp = prev.next
            prev.next = temp.next
            this.length --
            return temp
        }
    }
    reverse() {
        if (this.length <= 1) {
            return this
        }
        let cur = this.head
        this.head = this.tail
        this.tail = cur
        let prev = null
        let next
        for (let i = 0; i < this.length; i++) {
            next = cur.next
            cur.next = prev
            prev = cur
            cur = next
        }
        return this
    }
    print() {
        let node = this.head
        let arr = []
        for (let i=0; i<this.length; i++) {
            arr.push(node.val)
            node = node.next
        }
        console.log(arr, arr.length)
    }
}

let linkedList = new SinglyLinkedList()

console.log(linkedList.push(30))
console.log(linkedList.push(40))
console.log(linkedList.unshift(50))
linkedList.print()
linkedList.insert(1, 60)
linkedList.print()
linkedList.reverse()
linkedList.print()

////////////////////////////////////////////////////////////////////////////////////

class Node {
    constructor(val) {
        this.val = val
        this.prev = null
        this.next = null
    }
}

class DoublyLinkedList {
    constructor() {
        this.head = null
        this.tail = null
        this.length = 0
    }
    push(val) {
        let node = new Node(val)
        if (this.length === 0) {
            this.head = node
            this.tail = node
        } else {
            this.tail.next = node
            node.prev = this.tail
            this.tail = node
        }
        this.length ++
        return this
    }
    pop() {
        if (this.length === 0) {
            return null
        } else if (this.length === 1) {
            this.head = null
            this.tail = null
        }
        let node = this.tail
        this.tail = node.prev
        node.prev = null
        this.tail.next = null
        this.length --
        return node
    }
    shift() {
        if (this.length === 0) return undefined
        let temp = this.head
        if (this.length === 1) {
            this.head = null
            this.tail = null
        }
        this.head = temp.next
        this.head.prev = null
        temp.next = null
        this.length --
        return temp
    }
    unshift(val) {
        let node = new Node(val)
        if (this.length === 0) {
            this.head = node
            this.tail = node
        } else {
            this.head.prev = node
            node.next = this.head
            this.head = node
        }
        this.length ++
        return this
    }
    get(index) {
        let node
        if (index < 0 || index > this.length) {
            return null
        }
        if (index > this.length/2) {
            index = this.length - 1 - index
            node = this.tail
            while(index != 0) {
                node = node.next
                index --
            }
        } else {
            node = this.head
            while(index != 0) {
                node = node.next
                index --
            }
        }
        return node
    }
    set(index, val) {
        let node = this.get(index)
        if (node) {
            node.val = val
            return true
        }
        return false
    }
    insert(index, val) {
        if (index < 0 || index > this.length) {
            return false
        } else if (index === 0) {
            this.unshift(val)
        } else if (index === this.length) {
            this.push(val)
        } else {
            let prev = this.get(index - 1)
            let next = prev.next
            let newNode = new Node(val)
            prev.next = newNode
            newNode.prev = prev
            newNode.next = next
            next.prev = newNode
            this.length ++
        }
        return true
    }
    remove(index) {
        if (index < 0 || index >= this.length) {
            return null
        } else if (index === 0) {
            return this.shift()
        } else if (index === this.length - 1) {
            return this.pop()
        } else {
            let prev = this.get(index - 1)
            let next = prev.next.next
            let temp = prev.next
            prev.next = next
            next.prev = prev
            temp.next = null
            temp.prev = null
            this.length --
            return temp
        }
    }
    reverse() {
        if (this.length <= 1) {
            return this
        }
        let cur = this.head
        this.head = this.tail
        this.tail = cur
        let prev = null
        let next
        for (let i = 0; i < this.length; i++) {
            next = cur.next
            cur.next = prev
            if (prev != null) {
                prev.prev = cur
            }
            prev = cur
            cur = next
        }
        return this
    }
    print() {
        let node = this.head
        let arr = []
        for (let i=0;i<this.length;i++) {
            arr.push(node.val)
            node = node.next
        }
        console.log(arr)
    }
}

File: c_p_p/linkedLists.cpp
#include <iostream>

class Node
{
public:
    int val;
    Node *next;
    Node(int val) : val(val), next(nullptr) {}
};

class SingleLinkedList
{
private:
    Node *head;
    Node *tail;
    int length = 0;

public:
    bool push(int val)
    {
        if (head == nullptr)
        {
            head = new Node(val);
            tail = head;
            length++;
            return true;
        }
        Node *temp = new Node(val);
        tail->next = temp;
        tail = tail->next;
        length++;
        return true;
    }

    int pop()
    {
        if (head == nullptr)
        {
            std::cout << "empty" << std::endl;
            return 0;
        }
        if (head->next == nullptr)
        {
            int val = head->val;
            Node *temp = head;
            head = nullptr;
            delete temp;
            length--;
            return val;
        }
        Node *cur = head;
        while (cur->next->next != nullptr)
        {
            cur = cur->next;
        }
        int val = cur->next->val;
        Node *temp = cur->next;
        cur->next = nullptr;
        delete temp;
        length--;
        return val;
    }

    bool unshift(int val)
    {
        if (head == nullptr)
        {
            push(val);
            return true;
        }
        Node *temp = new Node(val);
        temp->next = head;
        head = temp;
        length++;
        if (head == tail)
            tail = tail->next;
        return true;
    }

    int shift()
    {
        if (head == nullptr)
        {
            std::cout << "empty" << std::endl;
            return 0;
        }
        Node *temp = head;
        int val = head->val;
        head = head->next;
        delete temp;
        length--;
        return val;
    }

    Node *get(int index)
    {
        if (index >= length || index < 0)
        {
            std::cout << "Out of bounds" << std::endl;
            return nullptr;
        }
        else if (index == length - 1)
        {
            return tail;
        }
        else if (index == 0)
        {
            return head;
        }
        Node *cur = head;
        for (int i{0}; i < length; i++)
        {
            if (i == index)
            {
                return cur;
            }
            cur = cur->next;
        }
        std::cout << "should never get here" << std::endl;
        return nullptr;
    }

    int take(int index)
    {
        if (index >= length || index < 0)
        {
            std::cout << "Out of bounds" << std::endl;
            return 0;
        }
        else if (index == length - 1)
        {
            pop();
        }
        else if (index == 0)
        {
            shift();
        }
        Node *cur = get(index - 1);
        int val = cur->next->val;
        Node *temp = cur->next;
        cur->next = cur->next->next;
        delete temp;
        return val;
    }
    void set(int index, int val)
    {
        if (index >= length || index < 0)
        {
            std::cout << "Out of bounds" << std::endl;
        }
        else if (index == length - 1)
        {
            tail->val = val;
        }
        else if (index == 0)
        {
            head->val = val;
        }
        get(index)->val = val;
    }
    bool insert(int index, int val)
    {
        if (index >= length || index < 0)
        {
            std::cout << "Out of bounds" << std::endl;
        }
        else if (index == length - 1)
        {
            return push(val);
        }
        else if (index == 0)
        {
            return unshift(val);
        }
        Node *cur = get(index - 1);
        Node *temp = new Node(val);
        temp->next = cur->next;
        cur->next = temp;
        return true;
    }

    bool reverse()
    {
        tail = head;
        Node *temp = head->next;
        Node *next = temp;
        head->next = nullptr;
        while (temp != nullptr)
        {
            next = next->next;
            temp->next = head;
            head = temp;
            temp = next;
        }
        delete temp, next;
        return true;
    }

    void print()
    {
        Node *cur = head;
        std::cout << "Print: Size = " << length << std::endl;
        while (cur != nullptr)
        {
            std::cout << cur->val << std::endl;
            cur = cur->next;
        }
    }
};